### PR TITLE
Adding ModalHeader component

### LINKED ISSUE #18032: [Ext] Create component: ModalHeader
### Description
Create a reusable UI component for `ModalHeader` according to the insight report

### References
[Modal Audit FigJam](https://www.figma.com/file/hxYqloYgmVcgsoiVqmGZ8K/Modal-Audit?node-id=0%3A1&t=WcWEbEnjEkGuOdXr-1)
[Figma component](https://www.figma.com/file/HKpPKij9V3TpsyMV1TpV7C/DS-Components?node-id=3180%3A13642)
[UI component guidelines](https://www.notion.so/UI-Components-5ccd2bf83eb8441892a0b72c0d8929e1)
[Styling guidelines](https://www.notion.so/Extension-Frontend-Engineering-Guide-b82ddb3e14004b5db3799a9b446294a9#49aa2ef7b7aa40178ff918043df71c34)
Testing guidelines: TBC

### Files needed
- `index.js`
- `index.scss`
- `modal-header.js`
- `modal-header.stories.js`
- `modal-header.test.js`
- `README.mdx`

### Technical details
TBC

### Acceptance criteria
- [ ] Uses color, typography, shadows design tokens 
- [ ] Uses semantic html
- [ ] PropTypes have descriptions
- [ ] Has storybook story with controls
- [ ] Has documentation in MDX
- [ ] Has unit tests and 90% coverage
- [ ] Works in Chrome and Firefox
- [ ] Performance tested: no unnecessary re-renders or other performance concerns

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## Explanation
Adding `ModalHeader` component according to insight report. As `ModalHeader` is almost a replica of `PopoverHeader` there are also some updates to improve `PopoverHeader` test coverage, stories and some small nits

* Fixes #18032 

### Before

`PopoverHeader` `children` prop control wasn't working as expected

https://user-images.githubusercontent.com/8112138/234163410-4066e8e5-e28d-488d-9d58-0cc18fdfa490.mov

### After

`PopoverHeader` adding some new stories and showing that the controls work

https://user-images.githubusercontent.com/8112138/234163660-be5d6ab2-b438-4679-a388-c779da787d6c.mov

Adding new `ModalHeader` component, stories, controls and documentation

https://user-images.githubusercontent.com/8112138/234163567-17ae9bb4-134a-4e68-b6be-91f388b9669f.mov

## Manual Testing Steps

- Go to the latest build of storybook on this PR
- Search `ModalHeader` 
- Check stories, controls, documentation and console
- Search `PopoverHeader`
- Check stories, controls, documentation and console

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/components/component-library/badge-wrapper/badge-wrapper.stories.tsx [hunk 1]
```diff
@@ -280,7 +280,7 @@ export const PositionObj = () => (
   </Box>
 );
 
-export const AnchorElementShape = () => (
+export const AnchorElementShape: ComponentStory<typeof BadgeWrapper> = () => (
   <Box display={DISPLAY.FLEX} gap={4}>
     <BadgeWrapper
       badge={
```
Comment (georgewrmarshall) [inline ui/components/component-library/badge-wrapper/badge-wrapper.stories.tsx:286]: Move to another PR

File: ui/components/component-library/button-icon/button-icon.d.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  /**
+   * The size of the ButtonIcon.
+   * Possible values could be 'Size.SM' 24px, 'Size.LG' 32px,
+   */
+  size?: typeof BUTTON_ICON_SIZES[number];
+}
+
+declare const ButtonIcon: React.FC<ButtonIconProps>;
+
+export default ButtonIcon;
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/button-icon.d.ts:51]: Move to another PR

File: ui/components/component-library/button-icon/button-icon.stories.tsx [hunk 3]
```diff
@@ -185,7 +190,7 @@ ColorStory.args = {
   color: Color.primaryDefault,
 };
 
-export const Disabled = (args) => (
+export const Disabled: ComponentStory<typeof ButtonIcon> = (args) => (
   <ButtonIcon {...args} iconName={ICON_NAMES.CLOSE} ariaLabel="close" />
 );
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/button-icon.stories.tsx:196]: Move to another PR

File: ui/components/component-library/header-base/__snapshots__/header-base.test.tsx.snap [hunk 4]
```diff
@@ -8,7 +8,7 @@ exports[`HeaderBase should render HeaderBase element correctly 1`] = `
     title="HeaderBase test"
   >
     <div
-      class="box mm-header-base__children box--flex-direction-row box--width-full"
+      class="box box--flex-direction-row box--width-full"
     >
       should render HeaderBase element correctly
     </div>
```
Comment (georgewrmarshall) [inline ui/components/component-library/header-base/__snapshots__/header-base.test.tsx.snap:14]: Removing unused classNames

File: ui/components/component-library/header-base/header-base.tsx [hunk 5]
```diff
@@ -98,7 +96,6 @@ export const HeaderBase: React.FC<HeaderBaseProps> = ({
         <Box
           display={DISPLAY.FLEX}
           justifyContent={JustifyContent.flexEnd}
-          className="mm-header-base__end-accessory"
           ref={endAccessoryRef}
           style={
             children
```
Comment (georgewrmarshall) [inline ui/components/component-library/header-base/header-base.tsx:101]: Removing unused classNames if engineers need a class they can assign one using the childrenWrapperProps

File: ui/components/component-library/header-base/header-base.tsx [hunk 6]
```diff
@@ -86,7 +85,6 @@ export const HeaderBase: React.FC<HeaderBaseProps> = ({
       )}
       {children && (
         <Box
-          className="mm-header-base__children"
           width={BLOCK_SIZES.FULL}
           style={getTitleStyles}
           {...childrenWrapperProps}
```
Comment (georgewrmarshall) [inline ui/components/component-library/header-base/header-base.tsx:90]: Could make this a bit more robust by doing something like this
```suggestion
          {...childrenWrapperProps}
          style={{...getTitleStyles, ...childrenWrapperProps?.style}
          
```

File: ui/components/component-library/modal-header/README.mdx [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+## Props
+
+The `ModalHeader` accepts all props below as well as all [Box](/docs/components-ui-box--default-story#props) component props
+
+<ArgsTable of={ModalHeader} />
+
+`ModalHeader` accepts all [Text](/docs/components-componentlibrary-text--default-story#props) component props
+
+<ArgsTable of={Text} />
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/README.mdx:23]: Add docs for other stories

File: ui/components/component-library/modal-header/__snapshots__/modal-header.test.tsx.snap [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    <div
+      class="box box--flex-direction-row box--width-full"
+    >
+      <h4
+        class="box mm-text mm-text--heading-sm mm-text--text-align-center box--flex-direction-row box--color-text-default"
+      >
+        modal header
+      </h4>
+    </div>
+  </div>
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/__snapshots__/modal-header.test.tsx.snap:17]: Seems like a lot of div wrappers if there aren't any buttons I wonder if we can rectify this? cc @garrettbear

File: ui/components/component-library/modal-header/modal-header.test.tsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  });
+  it('should render with react nodes inside the ModalHeader', () => {
+    const { getByText } = render(
+      <ModalHeader>
+        <span>modal header</span>
+      </ModalHeader>,
+    );
+    expect(getByText('modal header')).toBeDefined();
+  });
+});
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/modal-header.test.tsx:35]: Add more tests

File: ui/components/component-library/modal-header/modal-header.tsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+          <ButtonIcon
+            iconName={ICON_NAMES.CLOSE}
+            onClick={onClickCloseButton}
+            size={BUTTON_ICON_SIZES.SM}
+            {...closeButtonProps}
+          />
+        ))
+      }
+      {...props}
+    >
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/modal-header.tsx:56]: Hmm need some typescript whizz to help us create types for polymorphic types it would be cool to do something like changing the childrenWrapper to a `Text` component 
```
childrenWrapperProps={{
        as: Text,
        variant: TextVariant.headingSm,
      }}
```
cc @garrettbear

File: ui/components/component-library/text/text.d.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  ellipsis?: boolean;
+  as?: ValidTags;
+  className?: string;
+}
+
+export const Text: React.ForwardRefExoticComponent<
+  TextProps & React.RefAttributes<HTMLElement>
+>;
+
+export const getTextElementDefault: (variant: TextVariant) => ValidTags;
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.d.ts:53]: Move this to a separate PR

File: ui/components/component-library/text/text.stories.tsx [hunk 12]
```diff
@@ -290,7 +290,7 @@ export const As = (args) => (
   </>
 );
 
-export const TextDirection = (args) => (
+export const TextDirection: ComponentStory<typeof Text> = (args) => (
   <Box
     style={{ maxWidth: 300 }}
     display={DISPLAY.FLEX}
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.stories.tsx:296]: Move to a separate PR

File: ui/components/component-library/text/text.types.ts [hunk 13]
```diff
@@ -88,7 +90,7 @@ export interface TextProps extends BoxProps {
    * The text-align of the Text component. Should use the TEXT_ALIGN object from
    * ./ui/helpers/constants/design-system.js
    */
-  textAlign?: keyof typeof TEXT_ALIGN;
+  textAlign?: TextAlign;
   /**
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.types.ts:94]: Adding so TEXT_ALIGN object works. This will eventually be replaced when https://github.com/MetaMask/metamask-extension/issues/18444 is completed

File: ui/components/component-library/badge-wrapper/badge-wrapper.stories.tsx (hunk 1 above)
Comment (georgewrmarshall) [reply ui/components/component-library/badge-wrapper/badge-wrapper.stories.tsx]: done

File: ui/components/component-library/header-base/header-base.tsx (hunk 6 above)
Comment (georgewrmarshall) [reply ui/components/component-library/header-base/header-base.tsx]: done

File: ui/components/component-library/modal-header/__snapshots__/modal-header.test.tsx.snap (hunk 8 above)
Comment (georgewrmarshall) [reply ui/components/component-library/modal-header/__snapshots__/modal-header.test.tsx.snap]: Updated

File: ui/components/component-library/modal-header/modal-header.tsx (hunk 10 above)
Comment (georgewrmarshall) [reply ui/components/component-library/modal-header/modal-header.tsx:45]: Updated to make children wrapper a Text component

File: ui/components/component-library/button-icon/button-icon.d.ts [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+   */
+  href?: string;
+  /**
+   * The name of the icon to display. Should be one of ICON_NAMES
+   */
+  iconName: IconName;
+  /**
+   * iconProps accepts all the props from Icon
+   */
+  iconProps?: any;
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/button-icon.d.ts:43]: Update to use `IconProps` interface

File: ui/components/component-library/button-icon/button-icon.d.ts [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+  /**
+   * The size of the ButtonIcon.
+   * Possible values could be 'Size.SM' 24px, 'Size.LG' 32px,
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/button-icon.d.ts:46]: Update to use `BUTTON_ICON_SIZES`

File: ui/components/component-library/header-base/header-base.test.tsx [hunk 16]
```diff
@@ -58,4 +58,19 @@ describe('HeaderBase', () => {
...
+          'data-testid': 'children-wrapper',
+          style: { color: 'red' },
+        }}
+      >
+        HeaderBase children test
+      </HeaderBase>,
+    );
+    expect(getByTestId('children-wrapper')).toBeDefined();
+    expect(getByTestId('children-wrapper')).toHaveStyle(`color: red`);
+  });
```
Comment (georgewrmarshall) [inline ui/components/component-library/header-base/header-base.test.tsx:75]: Adding more test coverage

File: ui/components/component-library/modal-header/modal-header.types.ts [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+import type { ButtonIconProps } from '../button-icon/button-icon.types';
+import type { HeaderBaseProps } from '../header-base';
+
+export interface ModalHeaderProps extends HeaderBaseProps {
+  /**
+   * The contents within the ModalHeader positioned middle (popular for title use case)
+   */
+  children?: React.ReactNode;
+  /**
+   * Additional classNames to be added to the Modal Header component
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/modal-header.types.ts:11]:
```suggestion
   * Additional classNames to be added to the ModalHeader component
```

File: ui/components/component-library/popover-header/README.mdx [hunk 18]
```diff
@@ -3,7 +3,7 @@ import { PopoverHeader } from './popover-header';
 
 # PopoverHeader
 
-PopoverHeader is built on top of [HeaderBase](/docs/components-componentlibrary-headerbase--default-story) component with the most common use case of a back button in the startAccessory position, title, and close button in the endAccessory position.
+`PopoverHeader` handles the title, close button and back button for all Modal components. It is built on top of [HeaderBase](/docs/components-componentlibrary-headerbase--default-story) component
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/README.mdx:6]:
```suggestion
`PopoverHeader` handles the title, close button and back button for all `Popover` components. It is built on top of [HeaderBase](/docs/components-componentlibrary-headerbase--default-story) component
```

File: ui/components/component-library/popover-header/README.mdx [hunk 19]
```diff
@@ -17,16 +17,41 @@ The `PopoverHeader` accepts all props below as well as all [Box](/docs/ui-compon
 
 ### Children
 
-Wrapping string content in the `PopoverHeader` component will be rendered in the center of the header with the default title `Text` component.
+The title of the `PopoverHeader` component. Passing a `string` will render the content inside of a `Text` component. Passing any other type will render the content as is.
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/README.mdx:21]: Updating description to explain the logic a little more

File: ui/components/component-library/popover-header/README.mdx [hunk 20]
```diff
@@ -17,16 +17,41 @@ The `PopoverHeader` accepts all props below as well as all [Box](/docs/ui-compon
...
+    alignItems: AlignItems.center,
+    justifyContent: JustifyContent.center,
+  }}
+>
+  <AvatarAccount address="0x1234" />
+  <Text variant={TextVariant.headingSm} textAlign={TextAlign.Center}>
+    Custom header using multiple components
+  </Text>
+</PopoverHeader>
 ```
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/README.mdx:55]: Adding example of custom header

File: ui/components/component-library/popover-header/popover-header.stories.tsx [hunk 21]
```diff
@@ -24,42 +35,52 @@ export default {
 } as ComponentMeta<typeof PopoverHeader>;
 
 const Template: ComponentStory<typeof PopoverHeader> = (args) => {
-  return <PopoverHeader {...args}>PopoverHeader</PopoverHeader>;
+  return <PopoverHeader {...args} />;
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/popover-header.stories.tsx:38]: Fixes `children` control so it works

File: ui/components/component-library/popover-header/popover-header.stories.tsx [hunk 22]
```diff
@@ -24,42 +35,52 @@ export default {
...
+        alignItems: AlignItems.center,
+        justifyContent: JustifyContent.center,
+      }}
+    >
+      <AvatarAccount address="0x1234" />
+      <Text variant={TextVariant.headingSm} textAlign={TextAlign.Center}>
+        Custom header using multiple components
+      </Text>
+    </PopoverHeader>
+  </>
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/popover-header.stories.tsx:63]: Adding custom header example

File: ui/components/component-library/popover-header/popover-header.stories.tsx [hunk 23]
```diff
@@ -24,42 +35,52 @@ export default {
...
-    {...args}
-  >
-    EndAccessory
-  </PopoverHeader>
-);
+export const EndAccessory = Template.bind({});
+EndAccessory.args = {
+  children: 'EndAccessory demo',
+  endAccessory: <Button size={BUTTON_SIZES.SM}>Demo</Button>,
+};
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/popover-header.stories.tsx:86]: Fixing all `children` controls and using storybook way to repeat templates and utiles `.args` to reduce amount of code

File: ui/components/component-library/popover-header/popover-header.test.tsx [hunk 24]
```diff
@@ -14,24 +14,33 @@ describe('PopoverHeader', () => {
     expect(container).toMatchSnapshot();
   });
 
-  it('should render popover header title', () => {
+  it('should render popover header children as a string', () => {
     const { getByText } = render(
       <PopoverHeader data-testid="popover-header">
-        Popover Header Test
+        Modal header test
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/popover-header.test.tsx:20]:
```suggestion
        PopoverHeader test
```

File: ui/components/component-library/popover-header/popover-header.test.tsx [hunk 25]
```diff
@@ -14,24 +14,33 @@ describe('PopoverHeader', () => {
...
       </PopoverHeader>,
     );
-    expect(getByText('Popover Header Test')).toBeDefined();
+    expect(getByText('Modal header test')).toBeDefined();
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/popover-header.test.tsx:23]:
```suggestion
    expect(getByText('PopoverHeader test')).toBeDefined();
```

File: ui/components/component-library/popover-header/popover-header.test.tsx [hunk 26]
```diff
@@ -14,24 +14,33 @@ describe('PopoverHeader', () => {
...
+  });
+
+  it('should render popover header children as a node', () => {
+    const { getByText, getByTestId } = render(
+      <PopoverHeader data-testid="popover-header">
+        <div data-testid="div">Modal header test</div>
+      </PopoverHeader>,
+    );
+    expect(getByText('Modal header test')).toBeDefined();
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/popover-header.test.tsx:32]:
```suggestion
        <div data-testid="div">PopoverHeader test</div>
      </PopoverHeader>,
    );
    expect(getByText('PopoverHeader test')).toBeDefined();
```

File: ui/components/component-library/popover-header/popover-header.types.ts [hunk 27]
```diff
@@ -8,7 +8,7 @@ export interface PopoverHeaderProps extends HeaderBaseProps {
    */
   children?: React.ReactNode;
   /**
-   * Additional classNames to be added to the Popover component
+   * Additional classNames to be added to the Popover Header component
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/popover-header.types.ts:11]:
```suggestion
   * Additional classNames to be added to the PopoverHeader component
```

File: ui/components/component-library/header-base/header-base.tsx [hunk 28]
```diff
@@ -70,7 +70,6 @@ export const HeaderBase: React.FC<HeaderBaseProps> = ({
     >
       {startAccessory && (
         <Box
-          className="mm-header-base__start-accessory"
```
Comment (georgewrmarshall) [inline ui/components/component-library/header-base/header-base.tsx:73]: Removing classNames here as there are no associated styles and should encourage engineers to add their own className if required through `childrenWrapperProps` instead of using this one which could be subject to change.

File: ui/components/component-library/modal-header/README.mdx [hunk 29]
```diff
@@ -0,0 +1,123 @@
+import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';
+import { ModalHeader } from './modal-header';
+
+# ModalHeader
+
+`ModalHeader` handles the title, close button and back button for all Modal components. It is built on top of the [HeaderBase](/docs/components-componentlibrary-headerbase--default-story) component
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/README.mdx:6]:
```suggestion
`ModalHeader` handles the title, close button and back button for all Modal components. It is built on top of the [HeaderBase](/docs/components-componentlibrary-headerbase--default-story) component.
```

File: ui/components/component-library/modal-header/README.mdx [hunk 30]
```diff
@@ -0,0 +1,123 @@
...
+  <AvatarAccount address="0x1234" />
+  <Text variant={TextVariant.headingSm} textAlign={TextAlign.Center}>
+    Custom header using multiple components
+  </Text>
+</ModalHeader>
+```
+
+### onBack
+
+Use the onClick handler `onBack` prop to render the `ButtonIcon` back button in the startAccessory position.
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/README.mdx:59]:
```suggestion
Use the onClick handler `onBack` prop to render the `ButtonIcon` back button in the startAccessory position
```

File: ui/components/component-library/modal-header/README.mdx [hunk 31]
```diff
@@ -0,0 +1,123 @@
...
+
+Use the `backButtonProps` prop to pass additional props to the `ButtonIcon` back button.
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/README.mdx:61]:
```suggestion
Use the `backButtonProps` prop to pass additional props to the `ButtonIcon` back button
```

File: ui/components/component-library/modal-header/README.mdx [hunk 32]
```diff
@@ -0,0 +1,123 @@
...
+import { ModalHeader } from '../../component-library';
+
+<ModalHeader onBack={() => console.log('Back button click')}>
+  OnBack Demo
+</ModalHeader>;
+```
+
+### onClose
+
+Use the onClick handler `onClose` prop to render the `ButtonIcon` back button in the endAccessory position.
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/README.mdx:77]:
```suggestion
Use the onClick handler `onClose` prop to render the `ButtonIcon` back button in the endAccessory position
```

File: ui/components/component-library/modal-header/README.mdx [hunk 33]
```diff
@@ -0,0 +1,123 @@
...
+
+Use the `backButtonProps` prop to pass additional props to the `ButtonIcon` back button.
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/README.mdx:79]:
```suggestion
Use the `backButtonProps` prop to pass additional props to the `ButtonIcon` back button
```

File: ui/components/component-library/popover-header/popover-header.test.tsx [hunk 34]
```diff
@@ -6,32 +6,39 @@ import { PopoverHeader } from './popover-header';
...
+
+  it('should render popover header children as a node', () => {
+    const { getByText, getByTestId } = render(
+      <PopoverHeader data-testid="popover-header">
+        <div data-testid="div">PopoverHeader test</div>
       </PopoverHeader>,
     );
-    expect(getByText('Popover Header Test')).toBeDefined();
+    expect(getByText('PopoverHeader test')).toBeDefined();
+    expect(getByTestId('div')).toBeDefined();
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/popover-header.test.tsx:31]: Increasing test coverage for `PopoverHeader`

### Before
Screenshot 2023-04-25 at 8 52 54 AM

### After

Screenshot 2023-04-25 at 8 54 07 AM

**Review by NidhiKJha — APPROVED** [review]
LGTM 🚀

**Review by garrettbear — APPROVED** [review]
LGTM, not sure if you were making those changes with the `.` periods
